textattack/shared/attack.py

```diff
@@ -182,85 +182,64 @@ def attack_one(self, initial_result):
         """ 
         Gets examples from a dataset and tokenizes them.
 
         Args:
             dataset: An iterable of (text, ground_truth_output) pairs
-            num_examples (int): the number of examples to return
-            shuffle (:obj:`bool`, optional): Whether to shuffle the data
-            attack_n (bool): If `True`, returns `num_examples` non-skipped
-                examples. If `False`, returns `num_examples` total examples.
+            indices: An iterable of indices of the dataset that we want to attack. If None, attack all samples in dataset.
         
         Returns:
-            results (Iterable[Tuple[GoalFunctionResult, Boolean]]): a list of
-                objects containing (text, ground_truth_output, was_skipped)
+            results (Iterable[GoalFunctionResult]): an iterable of GoalFunctionResults of the original examples
         """
-        examples = []
-        n = 0
-
-        if shuffle:
-            random.shuffle(dataset.examples)
-
-        num_examples = num_examples or len(dataset)
-
-        if num_examples <= 0:
+        indices = indices if indices else deque(range(len(dataset)))
+        if not isinstance(indices, deque):
+            indices = deque(indices)
+        if not indices:
             return
             yield
 
-        for text, ground_truth_output in dataset:
-            tokenized_text = TokenizedText(text, self.goal_function.tokenizer)
-            self.goal_function.num_queries = 0
-            goal_function_result, _ = self.goal_function.get_result(
-                tokenized_text, ground_truth_output
-            )
-            # We can skip examples for which the goal is already succeeded,
-            # unless `attack_skippable_examples` is True.
-            if (not attack_skippable_examples) and (goal_function_result.succeeded):
-                if not attack_n:
-                    n += 1
-                # Store the true output on the goal function so that the
-                # SkippedAttackResult has the correct output, not the incorrect.
-                goal_function_result.output = ground_truth_output
-                yield (goal_function_result, True)
-            else:
-                n += 1
-                yield (goal_function_result, False)
-            if num_examples is not None and (n >= num_examples):
-                break
-
-    def attack_dataset(self, dataset, num_examples=None, shuffle=False, attack_n=False):
+        while indices:
+            i = indices.popleft()
+            try:
+                text, ground_truth_output = dataset[i]
+                tokenized_text = AttackedText(text)
+                self.goal_function.num_queries = 0
+                goal_function_result, _ = self.goal_function.get_result(
+                    tokenized_text, ground_truth_output
+                )
+                if goal_function_result.succeeded:
+                    # Store the true output on the goal function so that the
```

Comment:
Apparently this was already here but this is confusing to me. Why do we have to set the output? Feels like there should be a better way to do this. Maybe I can change it as part of my score function PR?

Reply:
Well that's for when model incorrectly predicts for the given output, so we display ground truth for skipped samples. I'm not sure if there's a better place to being doing this since it's the one place where we're handling `SkippedAttackResult`

Reply:
yes I agree @uvafan you should change this – every `AttackedText` object should have its ground truth output stored somewhere I think. Then we can print it to the CSV no matter what.